text/0036-ternary.md

```diff
@@ -0,0 +1,254 @@
+at once.
+
+We hope that `bee-ternary` will allow IOTA developers to more easily write code
+that works with ternary and allow the simplification of existing codebases
+through use of the API.
+
+Broadly, there are 3 main benefits that `bee-ternary` aims to provide
+over use-specific implementations of ternary manipulation code.
+
+- **Ergonomics**: The API should be trivial to use correctly and should
+naturally guide developers towards correct and efficient solutions.
+
+- **Features**: The API should provide a fundamental set of core features that
+can be used together to cover most developer requirements. Examples of such
+features include multiple encoding schemes, binary/ternary conversion,
+(de)serialization, and tryte string de/encoding.
+
+- **Performance**: The API should allow 'acceptably' efficient manipulation of
+ternary data. This is difficult to ruggedly define, and implementation details
+may change later, but broadly we intend to introduce an API that doesn't
+inherently inhibit high performance through poor design choices.
+
+# Detailed design
+
+`bee-ternary` is designed to be extensible and is built on top of a handful of
+fundamental abstraction types. Below we list the features of the API and
+summarize a few of the core API features.
+
+## Features
+
+- Efficient manipulation of ternary buffers (trits and trytes)
+- Multiple encoding schemes
+- Extensible design that allows it to sit on top of existing data structures,
+avoiding unnecessary allocation and copying
+- An array of utility functions to allow for easy manipulation of ternary data
+- Zero-cost conversion between trit and tryte formats
+
+## Key Types & Traits
+
+### `Btrit` and `Utrit`
+
+```rust
+enum Btrit { NegOne, Zero, PlusOne }
+enum Utrit { Zero, One, Two }
+```
+
+The crate supports both balanced and unbalanced trit representations, supported
+through two types, `Btrit` and `Utrit`. `Btrit` is the more common
+representation and so is the implicit default for most operations in the crate.
+
+### `Trits`
+
+```rust
+struct Trits<T: RawEncoding> { ... }
+```
+
+- Analogous to `str` or `[T]`
+- Unsized type, represents a buffer of trits of a specified encoding
+- Most commonly used from behind a reference (as with `&str` and `&[T]`),
+representing a 'trit slice'
+- Can be created from a variety of types such as`TritBuf` and `[i8]` with
+minimal overhead
+- Sub-slices can be created from trit slices at a per-trit level
+
+### `TritBuf`
+
+```rust
+struct TritBuf<T: RawEncodingBuf> { ... }
+```
+
+- Analogous to `String` or `Vec<T>`
+- Most common way to manipulate trits
+- Allows pushing, popping, and collecting trits
+- Implements `Deref<Target=Trits>` (in the same way that `Vec<T>` implements
+`Deref<[T]>`)
+
+### `RawEncoding`
+
+```rust
+trait RawEncoding { ... }
+```
+
+- Represents a raw trit buffer of some encoding
+- Common interface implemented by all trit encodings (`T1B1`, `T3B1`, `T5B1`,
```

Comment:
These are never actually defined, should we expand the `Encoding` a tiny bit and define these acronyms ?

Reply:
Will do